### PR TITLE
make print_with_color emit starting and terminating ANSI sequences on each line

### PR DESCRIPTION
**Pull request by KristofferC**
This is useful if you for example want to print something before each line of colored text.

For example:

```
function boxify(io, str)
	lines = split(str, '\n')
	for (i, line) in enumerate(lines)
		i == 1 || print(io, '\n')
		if length(lines) == 1
			print(io, "[ ")
		elseif i == 1
			print(io, "┌ ")
		elseif i == length(lines)
			print(io, "└ ")
		else
			print(io, "│ ")
		end
		print(io, line)
	end
end

str = """
This is a string
that has many lines
we will print it
in color"""

io = IOBuffer(); print_with_color(:red, IOContext(io, :color => true), str)
str_colored = String(take!(io));
boxify(STDOUT, str_colored)
```

Before this PR this will show

screen shot 2017-12-27 at 23 53 31

now it shows:

screen shot 2017-12-27 at 23 52 55

### CODE REVIEW

File: base/util.jl [hunk 1]
```diff
@@ -320,13 +320,21 @@ end
...
     finally
-        iscolor && color != :nothing && print(buf, get(disable_text_style, color, text_colors[:default]))
-        iscolor && (bold || color == :bold) && print(buf, disable_text_style[:bold])
-        print(io, String(take!(buf)))
+        lines = split(String(take!(buf)), '\n')
+        first = true
+        iob = IOBuffer()
+        for line in lines
+            first || print(iob, '\n')
+            first = false
```
Comment (KristofferC) [inline base/util.jl]: Perhaps check if line is empty here and `continue` do not write sequences on empty lines.

Comment (KristofferC) [reply base/util.jl]: Done

File: base/util.jl [hunk 2]
```diff
@@ -320,13 +320,22 @@ end
...
+            isempty(line) && continue
+            iscolor && bold && print(iob, text_colors[:bold])
+            iscolor && print(iob, get(text_colors, color, color_normal))
+            print(iob, line)
+            iscolor && color != :nothing && print(iob, get(disable_text_style, color, text_colors[:default]))
+            iscolor && (bold || color == :bold) && print(iob, disable_text_style[:bold])
```
Comment (vtjnash) [inline base/util.jl]: it seems silly to copy `line` to `iob` one at a time if `!iscolor`

**Review by vtjnash — COMMENTED** [review]
This does sound like a good idea. It would also help any time formatting gets messed up in the middle of a line, by ensuring that the formatting gets reset for the next line.

File: base/util.jl (hunk 2 above)
Comment (KristofferC) [reply base/util.jl]: Updated with 

```
if !iscolor
     print(io, str)
else
    ...
```

**Comment by KristofferC** [comment]
Good to go?

File: base/util.jl [hunk 3]
```diff
@@ -320,13 +320,27 @@ end
...
-        iscolor && color != :nothing && print(buf, get(disable_text_style, color, text_colors[:default]))
-        iscolor && (bold || color == :bold) && print(buf, disable_text_style[:bold])
-        print(io, String(take!(buf)))
+        str = String(take!(buf))
+        if !iscolor
+            print(io, str)
+        else
+            lines = split(str, '\n')
+            first = true
+            iob = IOBuffer()
```
Comment (stevengj) [inline base/util.jl]: Can we re-use `buf` here since it was just emptied?

File: base/util.jl [hunk 4]
```diff
@@ -320,13 +320,27 @@ end
...
+            for line in lines
+                first || print(iob, '\n')
+                first = false
+                isempty(line) && continue
+                iscolor && bold && print(iob, text_colors[:bold])
```
Comment (stevengj) [inline base/util.jl]: `iscolor` is always `true` in this loop, no?

File: base/util.jl [hunk 5]
```diff
@@ -320,13 +320,27 @@ end
...
+                iscolor && print(iob, get(text_colors, color, color_normal))
+                print(iob, line)
+                iscolor && color != :nothing && print(iob, get(disable_text_style, color, text_colors[:default]))
+                iscolor && (bold || color == :bold) && print(iob, disable_text_style[:bold])
```
Comment (stevengj) [inline base/util.jl]: Won't the `color == :bold` case be handled by the previous line?  It seems like you are printing `disable_text_style[:bold]` twice for `color == :bold`.

File: base/util.jl (hunk 3 above)
Comment (KristofferC) [reply base/util.jl]: Yes, done.

File: base/util.jl (hunk 4 above)
Comment (KristofferC) [reply base/util.jl]: Indeed, removed the checks.

File: base/util.jl (hunk 5 above)
Comment (KristofferC) [reply base/util.jl]: This is true. Fixed. Also made it not double print when passing both `:color = bold` and `bold = true`.

**Comment by KristofferC** [comment]
Updated from @stevengj's review

File: base/util.jl [hunk 6]
```diff
@@ -320,13 +320,26 @@ end
...
+        if !iscolor
+            print(io, str)
+        else
+            lines = split(str, '\n')
+            first = true
+            for line in lines
+                first || print(buf, '\n')
+                first = false
+                isempty(line) && continue
+                bold && color != :bold && print(buf, text_colors[:bold])
```
Comment (stevengj) [inline base/util.jl]: Seems better to just add a
```
bold && color == :bold && (color = :nothing)
```
line above.

Comment (KristofferC) [reply base/util.jl]: Alright, did this and shoved the dictionary lookups outside the loop.

File: base/util.jl [hunk 7]
```diff
@@ -326,17 +326,18 @@ function with_output_color(f::Function, color::Union{Int, Symbol}, io::IO, args.
         if !iscolor
             print(io, str)
         else
+            bold && color == :bold && (color = :nothing)
+            enable_ansi  = get(text_colors, color, text_colors[:default]) *
+                               (bold ? text_colors[:bold] : "")
+            disable_ansi = get(disable_text_style, color, text_colors[:default]) *
+                               (bold ? disable_text_style[:bold] : "")
```
Comment (stevengj) [inline base/util.jl]: Does it matter if you don't end in the opposite order?  i.e. if you do `red*bold` does it need to end with `disablebold*defaultcolor`?

File: base/util.jl [hunk 8]
```diff
@@ -326,17 +326,18 @@ function with_output_color(f::Function, color::Union{Int, Symbol}, io::IO, args.
...
             lines = split(str, '\n')
             first = true
             for line in lines
```
Comment (stevengj) [inline base/util.jl]: `for line in split(str, '\n')` would make it clearer that the `lines` variable is not used elsewhere.

File: base/util.jl (hunk 7 above)
Comment (KristofferC) [reply base/util.jl]: It doesn't matter but changed it anyway. It seems nicer.

File: base/util.jl (hunk 8 above)
Comment (KristofferC) [reply base/util.jl]: Done

**Review by stevengj — APPROVED** [review]

**Comment by c42f** [comment]
Oh ho, very nice :+1:

**Pull request merged**